src/lib.rs

```diff
@@ -409,47 +435,42 @@ where
 
     fn next(&mut self) -> Option<Self::Item> {
         let mut bump = true; // Request advance of next iterator.
-        for (iter, word) in self.iters.iter_mut() {
-            if bump || word.is_none() {
-                match iter.next() {
-                    None => {
-                        // This shouldn't happen because we expect the iterators
-                        // to cycle. However, if it does, we're definitely done.
-                        return None;
-                    }
-                    Some(None) => {
-                        // This is the cycle end marker. We want to get another
-                        // new word from this iterator, and advance the *next*
-                        // iterator too.
-                        match iter.next() {
-                            None => return None,
-                            Some(None) => return None,
-                            Some(s) => *word = s,
-                        }
-                        bump = true
-                    }
-                    Some(s) => {
-                        // We have a new word from this iterator, so we do not
-                        // yet need to advance the next iterator.
-                        *word = s;
-                        bump = false
-                    }
+        for (iter, word) in &mut self.iters {
+            if !bump && word.is_some() {
+                continue;
+            }
+
+            match iter.next()? {
+                None => {
+                    // This is the cycle end marker. We want to get another
+                    // new word from this iterator, and advance the *next*
+                    // iterator too.
+                    let new_word = iter.next().flatten()?;
+                    *word = Some(new_word);
+                    bump = true
+                }
+                Some(s) => {
+                    // We have a new word from this iterator, so we do not
+                    // yet need to advance the next iterator.
+                    *word = Some(s);
+                    bump = false
```

Comment:
I'm not a huge fan of this more imperative style. I prefer the more functional (term used loosely) style of the original. I suppose this is more idiomatic Rust, but to me it's harder to understand. I'm totally biased of course 😁 Anyway, unless there are clear advantages to this style that I've missed – maybe it compiles to something way better – I'd prefer to stick with the original.

Reply:
I understand you only dislike the short circuiting if blocks, not the `.flatten()` and `?` usage?

Reply:
@allenap I'm not sure that last question reached you, I'm wondering if you want this whole part of `fn next()` code reverted, or just the particular `if ... { continue; }` block?